File: src/ee/core/internal/IMessageBridge.swift
//
//  IMessageBridge.swift
//  Pods
//
//  Created by eps on 6/7/20.
//

import Foundation

public typealias MessageHandler = (_ message: String) -> String
public typealias AsyncMessageResolver = (_ message: String) -> Void
public typealias AsyncMessageHandler =
    (_ message: String, _ resolver: @escaping AsyncMessageResolver) -> Void

@objc(EEIMessageBridge)
public protocol IMessageBridge {
    /// Registers a new handler to receive messages from C++.
    /// @param tag The unique tag of the handler.
    /// @param handler The handler.
    /// @return Whether the registration was successful.
    @discardableResult
    func registerHandler(_ tag: String,
                         _ handler: @escaping MessageHandler) -> Bool

    /// Registers a new async handler to receive messages from C++.
    /// @param tag The unique tag of the handler.
    /// @param handler The handler.
    /// @return Whether the registration was successful.
    @discardableResult
    func registerAsyncHandler(_ tag: String,
                              _ handler: @escaping AsyncMessageHandler) -> Bool

    /// Deregisters an existing handler not to receive messages from C++.
    /// @param tag The unique tag of the handler.
    /// @return Whether the deregistration was successful.
    @discardableResult
    func deregisterHandler(_ tag: String) -> Bool

    /// Calls a handler from C++ without a message.
    /// @param tag The unique tag of the handler.
    /// @return Reply message from C++.
    @discardableResult
    func callCpp(_ tag: String) -> String

    /// Calls a handler from C++ with a message.
    /// @param tag The unique tag of the handler.
    /// @param message The message.
    /// @return Reply message from C++.
    @discardableResult
    func callCpp(_ tag: String, _ message: String) -> String
}

File: proj.android_studio/ee-x-core/src/main/java/com/ee/IMessageBridge.kt
package com.ee

import androidx.annotation.AnyThread

typealias MessageHandler = (message: String) -> String
typealias AsyncMessageHandler = suspend (message: String) -> String

/**
 * Created by eps on 3/20/18.
 */
interface IMessageBridge {
    /**
     * Registers a new handler to receive messages from C++.
     *
     * @param tag     The unique tag of the handler.
     * @param handler The handler.
     * @return Whether the registration was successful.
     */
    @AnyThread
    fun registerHandler(tag: String, handler: MessageHandler): Boolean

    /**
     * Registers a new async handler to receive messages from C++.
     *
     * @param tag     The unique tag of the handler.
     * @param handler The handler.
     * @return Whether the registration was successful.
     */
    @AnyThread
    fun registerAsyncHandler(tag: String, handler: AsyncMessageHandler): Boolean

    /**
     * Deregisters an existing handler not to receive messages from C++.
     *
     * @param tag The unique
     * @return Whether the deregistration was successful.
     */
    @AnyThread
    fun deregisterHandler(tag: String): Boolean

    /**
     * Calls a handler from Java with a message.
     *
     * @param tag     The unique tag of the handler.
     * @param message The message.
     * @return Reply message from Java.
     */
    @AnyThread
    fun call(tag: String, message: String): String

    /**
     * Calls a handler from C++ without a message.
     *
     * @param tag The unique tag of the handler.
     * @return Reply message from c++.
     */
    @AnyThread
    fun callCpp(tag: String): String

    /**
     * Calls a handler from C++ with a message.
     *
     * @param tag     The unique tag of the handler.
     * @param message The message.
     * @return Reply message from C++.
     */
    @AnyThread
    fun callCpp(tag: String, message: String): String
}